flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java

```diff
@@ -355,13 +359,53 @@ public static RpcService createRpcService(
 				taskManagerHostname, taskManagerAddress.getHostAddress());
 		}
 
-		final int rpcPort = configuration.getInteger(ConfigConstants.TASK_MANAGER_IPC_PORT_KEY, 0);
+		final String portRangeDefinition = configuration.getString(TaskManagerOptions.RPC_PORT, "0");
 
-		checkState(rpcPort >= 0 && rpcPort <= 65535, "Invalid value for " +
-				"'%s' (port for the TaskManager actor system) : %d - Leave config parameter empty or " +
-				"use 0 to let the system choose port automatically.",
-			ConfigConstants.TASK_MANAGER_IPC_PORT_KEY, rpcPort);
+		// parse port range definition and create port iterator
+		Iterator<Integer> portsIterator;
+		try {
+			portsIterator = NetUtils.getPortRangeFromString(portRangeDefinition);
+		} catch (Exception e) {
+			throw new IllegalArgumentException("Invalid port range definition: " + portRangeDefinition);
+		}
+
+		while (portsIterator.hasNext()) {
+			// first, we check if the port is available by opening a socket
+			// if the actor system fails to start on the port, we try further
+			ServerSocket availableSocket = NetUtils.createSocketFromPorts(
```

Comment:
I guess you took this code from `BootstrapTools#startActorSystem()`, but I'm wondering why we don't pass the port directly to `createRpcService`.

Reply:
@zentol the document always states it support the port range : https://ci.apache.org/projects/flink/flink-docs-master/ops/config.html#taskmanager-rpc-port

Reply:
and we also has the requirement about this feature

Reply:
that's...not my point.

Currently:
1. You take a port from the range,
2. create a socket to check its availability,
2.a if success, goto 3
2.b if failure, goto 1
3. close the socket,
4. pass port to `createRpcService`
4.a if success, exit
4.b if failure, goto 1

So, why do step 2 and 3 at all?

Reply:
it pick a available port and check it is not be applied. these logic is refered to the exists logic. And the jm, tm also support the port range.

Reply:
I know you're re-using existing code. That doesn't mean the code is perfect and shouldn't be changed. 

The code explicitly says that the actor system initialization can also fail with a `BindException`:
```
try {
	return startActorSystem(configuration, listeningAddress, port, logger);
}
catch (Exception e) {
	// we can continue to try if this contains a netty channel exception
	Throwable cause = e.getCause();
	if (!(cause instanceof org.jboss.netty.channel.ChannelException ||
			cause instanceof java.net.BindException)) {
		throw e;
	} // else fall through the loop and try the next port
}
```

This means that the check with the socket is redundant.